zhaquirks/tuya/builder/__init__.py

```diff
@@ -0,0 +1,513 @@
+            | foundation.ZCLAttributeAccess.Report,
+        )
+        self.binary_sensor(
+            attribute_name=attribute_name,
+            cluster_id=TUYA_CLUSTER_ID,
+            endpoint_id=endpoint_id,
+            device_class=device_class,
+            initially_disabled=initially_disabled,
+            attribute_initialized_from_cache=attribute_initialized_from_cache,
+            translation_key=translation_key,
+            fallback_name=fallback_name,
+        )
+
+        return self
+
+    def tuya_sensor(
+        self,
+        dp_id: int,
+        attribute_name: str,
+        type: type,
+        converter: Optional[Callable[[Any,], Any,]] = None,
+        dp_converter: Optional[Callable[[Any,], Any,]] = None,
+        endpoint_id: int = 1,
+        divisor: int = 1,
+        multiplier: int = 1,
+        entity_type: EntityType = EntityType.STANDARD,
+        device_class: SensorDeviceClass | None = None,
+        state_class: SensorStateClass | None = None,
+        unit: str | None = None,
+        initially_disabled: bool = False,
+        attribute_initialized_from_cache: bool = True,
+        translation_key: str | None = None,
+        fallback_name: str | None = None,
+    ) -> QuirkBuilder:  # fmt: skip
+        """Add an EntityMetadata containing ZCLSensorMetadata and return self.
+
+        This method allows exposing a sensor entity in Home Assistant.
+        """
+
+        self.tuya_dp_attribute(
+            dp_id=dp_id,
+            attribute_name=attribute_name,
+            type=type,
+            converter=converter,
+            dp_converter=dp_converter,
+            access=foundation.ZCLAttributeAccess.Read
+            | foundation.ZCLAttributeAccess.Report,
+        )
+        self.sensor(
+            attribute_name=attribute_name,
+            cluster_id=TUYA_CLUSTER_ID,
+            endpoint_id=endpoint_id,
+            divisor=divisor,
+            multiplier=multiplier,
+            entity_type=entity_type,
+            device_class=device_class,
+            state_class=state_class,
+            unit=unit,
+            initially_disabled=initially_disabled,
+            attribute_initialized_from_cache=attribute_initialized_from_cache,
+            translation_key=translation_key,
+            fallback_name=fallback_name,
+        )
+
+        return self
+
+    def add_to_registry(self) -> QuirksV2RegistryEntry:
+        """Build the quirks v2 registry entry."""
+
+        class NewAttributeDefs(TuyaMCUCluster.AttributeDefs):
+            """Attribute Definitions."""
+
+        for attr in self.AttributeDefs:
+            setattr(NewAttributeDefs, attr.name, attr)
+
+        class TuyaReplacementCluster(TuyaMCUCluster):
+            """Replacement Tuya Cluster."""
+
+            data_point_handlers: dict[int, str]
+            dp_to_attribute: dict[int, DPToAttributeMapping]
+
+            class AttributeDefs(NewAttributeDefs):
+                """Attribute Definitions."""
```

Comment:
Is the `AttributeDefs` stuff done like this, so we don't accidentally modify the "global class"?
Couldn't we create an `AttributeDefs` instance in the ``__init__`` of the quirk builder and just set the attributes there? Or am I missing something?

Reply:
AttributeDefs are just classes… there are no instances

Reply:
We're "dynamically modifying" `AttributeDefs` in `tuya_attribute`. I was trying to say/ask if this affects "all `AttributeDefs`" used in quirks created with the builder (which it should not) or if there's a specific `AttributeDefs` class/"instance" assigned to every quirk built with the builder (which should be the case).

Reply:
I was thinking if we need something similar to where `types.new_class` used here in zigpy, as we don't create an `AttributeDefs` class with attributes that are the same for all with the Tuya builder: https://github.com/zigpy/zigpy/blob/dfeb958b1f737b0a6c9dedbdb7fc169a188a516e/zigpy/zcl/__init__.py#L136-L143

Reply:
> I was thinking if we need something similar to where `types.new_class` is used here in zigpy, as we don't create an `AttributeDefs` class with attributes that are the same for all quirks created with the Tuya builder: https://github.com/zigpy/zigpy/blob/dfeb958b1f737b0a6c9dedbdb7fc169a188a516e/zigpy/zcl/__init__.py#L136-L143

I think this is the correct way to do it

Reply:
That said… these really are for making code clearer / easier to write…. Do we need the AttributeDefs here or is the dict syntax sufficient?

Reply:
I was in a hurry and missed a commit, this should be resolved now. While this seems odd, it's the only way I found that actually worked. Setting the attribute on `TuyaReplacementCluster` directly or replacing `AttributeDefs` similar to zigpy results in the attributes being dropped when passed to `replace`.

Reply:
> Do we need the AttributeDefs here or is the dict syntax sufficient?

(assuming "dict syntax" is just talking about the old-style zigpy `attributes` definition with a dict)
I was under the impression we no longer wanted to include them if possible. Although it's unlikely zigpy will drop them in the near future.. 😄 

If it's easily possible, I'd prefer to use the new-style `AttributeDefs`. We can also fall back to the old-style `attributes` otherwise (at least with the new `ZCLAttributeDef` objects).
I guess for a case like this, using `AttributeDefs` is more difficult than just using the dict style.

---

If we want to use new-style `AttributeDefs`, mind trying to see if something like this works? (replace the whole `add_to_registry` registry method with below).
Not saying this is the best solution, just curious if this works..

```python
    def add_to_registry(self) -> QuirksV2RegistryEntry:
        """Build the quirks v2 registry entry."""

        class TuyaReplacementCluster(TuyaMCUCluster):
            """Replacement Tuya Cluster."""

            data_point_handlers: dict[int, str]
            dp_to_attribute: dict[int, DPToAttributeMapping]

            def __init_subclass__(cls) -> None:
                cls.AttributeDefs = types.new_class(
                    name="AttributeDefs",
                    bases=(TuyaMCUCluster.AttributeDefs,),
                )
    
                for attr in self.new_attributes:
                    setattr(cls.AttributeDefs, attr.name, attr)

            async def write_attributes(self, attributes, manufacturer=None):
                """Overwrite to force manufacturer code."""
                return await super().write_attributes(
                    attributes, manufacturer=foundation.ZCLHeader.NO_MANUFACTURER_ID
                )
```

Reply:
Ok, so, apparently that doesn't seem to work. The current way in this PR works at least.

I've thought about it again and it's preferable to use `AttributeDefs` if possible.
If the base clusters (`TuyaMCUCluster` and so on) use `AttributeDefs`, we also have to. If they use the `attributes` dict, we also have to. 
Since `AttributeDefs` is the preferred way to go, we should use that here and convert the other clusters to `AttributeDefs`.

I'll try and take another stab at maybe slightly changing/improving what we have now (regarding `AttributeDefs`), but the current implementation works already..